### PR TITLE
Add BulkUpload history components

### LINKED ISSUE #413: Create upload history sidebar components
Users will be able to view a history of their bulk uploads (see #365 for full context):

Upload History

We will need three states: Processing, Error, and Complete.

- Processing's icon should be a rotating animation
- Relative times (e.g., "2 minutes ago") should be [actual `<time>` elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time)
- File size should be reported in nearest sensible unit (no decimals for KB; one decimal for MB; two decimals for GB)
- Error state does not currently report detailed error state; I'm still mulling on that

For the scope of this issue, we can assume what's shown here captures the full component need. Future features (like the ability to click on components to load detail in a panel) will come in separate/later issues.

### LINKED ISSUE #414: Populate file history sidebar
Using the components from #413, populate the file history sidebar, sorted newest (top) to oldest:

Upload sidebar

### LINKED ISSUE #492: Add call-to-action to BulkUpload "failed" message
Until detailed error reporting is completed, supplement "Failed [time] ago" message in `BulkUpload` history sidebar component with a call to action: "Failed [time] ago — email info@philanthropydatacommons.org for help".

### PR DESCRIPTION
**Pull request by reefdog**
This PR adds the BulkUpload history components and wires up the history to the API.

This is an incomplete PR but I'm hopping on a plane. Hopefully this establishes enough patterns and styling for you to continue from here! Still clearly TODO:

- [x] Demonstrate this in a panel
- [x] Humanize the byte size (1KB, 2MB, 1.1GB, etc.)
- [x] Humanize the time format
   - [x] Wrap the time in a `<time>` element with `datetime` attribute
- [x] Adapt to the real BulkUpload object attributes

**Update, January 19:**

Since writing the above, here's what's happened: Jason took over this branch, loaded BulkUpload history from the API, and wired up the components. (He also added file uploading on top of it, but that will be in a separate PR.) I collected that work into commits and added the humanized times and file sizes.

**Testing:**
- Check out the Storybook components
- Locally, create some BulkUpload history in your database (apologies, uploading itself will be in a forthcoming PR)
- Make sure the sidebar displays them correctly!

Note that the uploader itself is not included in this PR, and the styling/behavior of the components does not yet include the changes suggested in #477.

Closes #413
Closes #414
Closes #492

### CODE REVIEW

**Comment by reefdog** [comment]
@jasonaowen I’ve made some further refinements to these components but am having trouble getting my laptop on to the Istanbul airport WiFi, so I’ll have to wait to push them until I’m at my final spot late Thursday night your time.

**Comment by jasonaowen** [comment]
@reefdog how's your internet connection looking? Should I hold off on this or go with what's here?

File: src/interfaces/BulkUpload.ts
Comment (jasonaowen) [inline src/interfaces/BulkUpload.ts:4]: Hm, it looks like we are not storing the file size in the database at the moment, so the API is not returning it. That seems like something we could reasonably cache, but maybe not in this first iteration? cc @slifty

File: src/interfaces/BulkUpload.ts
Comment (jasonaowen) [inline src/interfaces/BulkUpload.ts:1]: I think I'm going to move this into `src/pdc-api.ts` as `ApiBulkUpload`.

File: src/components/BulkUploadListItem.tsx
Comment (jasonaowen) [inline src/components/BulkUploadListItem.tsx:53]: @slifty this makes me think we want the `bulk_uploads` table to have a `finished_at` or so!

File: src/components/BulkUploadListItem.tsx
Comment (jasonaowen) [inline src/components/BulkUploadListItem.tsx:16]: Here's the full list from https://github.com/PhilanthropyDataCommons/service/blob/eeb035f4a71cedf7fdeb5226b770349df853272a/src/types/BulkUpload.ts#L5-L9 :

- `pending`
- `in_progress`
- `completed`
- `failed`
- `canceled`

The `canceled` status is planned but not implemented, and the `pending` and `in_progress` can probably both be considered `processing` for the sake of the UI.

File: src/components/BulkUploadListItem.tsx
Comment (jasonaowen) [inline src/components/BulkUploadListItem.tsx:27]: I'm probably going to inline this to `['bulk-upload-list-item', status].join(' ')` or so.

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx]: Ah, that brings up a good note… I just YOLO'ed `createdAt` not knowing what the actual best timestamp attribute would be and meant to call attention to it. Thanks for noticing, and yes, any more appropriate stamp is fine.

I do think if we only show one timestamp, reporting to the user when _they_ took the action is preferable. But there are other good reasons for showing `finished_at`, including (probably) closer sync with the `created_at` timestamps of individual rows. In any case I doubt the differences will be material and am not at all insistent on one way or the other.

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx:30]: Yeah, I was considering doing the same, but wasn't sure how the status normalizing would work out / if we'd need to do more `switch`ing here. Feel free to do that.

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx]: Oh excellent, I almost guessed correctly! Thanks for surfacing those.

1. Is `canceled` planned for MVP? If so, I'll add styling for it. What does it mean? cc @slifty 
2. You're right that from our end user perspective, `pending` and `in_progress` are the same. (At least, going on my assumption about what they mean. Pending is received and in the queue, but not yet started; in progress has actually been started?)

Comment (reefdog) [reply src/components/BulkUploadListItem.tsx]: I'm gonna update based on this list and push again.

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx]: Oh, wait. I just realized the "Failed" status is written as if it's reporting when the failure occurred. Hmm.

Okay I've thought myself into thinking that for pending/in_progress we should report the start time (or something close to upload time) and for the other statuses we should report the end time (or something close). I don't think this precision is vital to the UX, but if it's easy enough to accomplish, then it'd be good.

In any case, I don't want to the blocker on that, so go with whatever you and @slifty think makes sense!

**Comment by reefdog** [comment]
@jasonaowen Thanks for the review notes! I just pushed my earlier changes from the airport (sorry!) and a few new ones based on your comments above, including the correct status values. I'll be offline for a while again, so please have full freedom over this!

File: src/interfaces/BulkUpload.ts
Comment (slifty) [reply src/interfaces/BulkUpload.ts:4]: Got it -- I'd vote we make an issue for it so we can implement it the Right Way after the demo.

File: src/components/BulkUploadListItem.tsx
Comment (slifty) [reply src/components/BulkUploadListItem.tsx]: For the immediate term if we could work with just the `createdAt` value that is preferred, I agree we should aspire to include more information before we consider the feature complete.  (this also gives us enough time to think about whether it would be better to have that more robust "status reporting" / "job history" feature which would provide the information in a more flexible / granular way along with the other desired metadata)

File: src/components/BulkUploadListItem.tsx
Comment (slifty) [reply src/components/BulkUploadListItem.tsx]: Just confirmation that we can skip canceled (in future the front end could add a "cancel" button for items that are pending, if you wanted.)

File: src/interfaces/BulkUpload.ts
Comment (jasonaowen) [reply src/interfaces/BulkUpload.ts:4]: https://github.com/PhilanthropyDataCommons/service/issues/651

File: src/components/BulkUploadListItem.tsx
Comment (jasonaowen) [reply src/components/BulkUploadListItem.tsx]: https://github.com/PhilanthropyDataCommons/service/issues/652

File: src/components/BulkUploadListItem.tsx
Comment (jasonaowen) [reply src/components/BulkUploadListItem.tsx]: > Pending is received and in the queue, but not yet started; in progress has actually been started

Correct!

**Comment by jasonaowen** [comment]
Rebased onto `main` after #432 and to prompt a deploy preview now that the base branch is `main`.

**Comment by jasonaowen** [comment]
Looking good! I still have a little bit left to clean up, but it's working locally:

Screen Shot 2023-12-04 at 16 10 28

**Comment by reefdog** [comment]
Huzzah! Clearly I'll need to do a further styling pass for situations like the wrapping (where the bullet delimiter appears at the beginning of the last line, not the end of the previous), and getting the spinner to be the same size as the icon, etc. But this is great!

**Comment by slifty** [comment]
I wanna give a plug for differentiating between "IN_PROGRESS" and "PENDING"

Something might be pending for a long time and that means something different from being in progress for a long time.  I think it would be helpful in debugging / user stories for there to be a distinction.

**Comment by reefdog** [comment]
I would describe them in the UI as “Waiting to be processed” and then “Processing”. (I think you picked fine names for the backend — better for that purpose, even — but they don’t translate intuitively IMO.)

I don’t mind, though I’d probably style them exactly the same and just change the bit of text in that metadata section.

**Comment by reefdog** [comment]
@slifty As is typical, I think I have an even **better** idea! Forthcoming…

**Comment by reefdog** [comment]
@slifty Take a look! https://deploy-preview-443--philanthropy-data-commons-viewer.netlify.app/storybook/?path=/docs/stories-bulkuploadlistitem--docs#default

**Comment by slifty** [comment]
@reefdog wow! It's perfect! And to think it's all thanks to me!

(this looks great -- I have no notes)

**Comment by reefdog** [comment]
@jasonaowen Relative time added to bulk upload history components. Should be visible in the Storybook deploy preview shortly. With that, I hand this back to you!

**Comment by reefdog** [comment]
@slifty This PR is ready for review! Very sorry that it basically includes bulk upload _history_ without actually supporting the uploading. 😬 So you'll have to review code + Storybook components, but if you want to see it in practice you'll need to locally stuff some bulk upload history into your database manually.

File: src/components/BulkUploadListItem.tsx
Comment (slifty) [inline src/components/BulkUploadListItem.tsx:98]: Another instance of array index being used as a key; does `datum` have any type of identifier?

File: src/utils/files.ts
Comment (slifty) [inline src/utils/files.ts:9]: It might be good to introduce this method in the same commit it is used (as written this commit just adds dead code, although of course I assume it's about to get used somewhere)

Comment (slifty) [reply src/utils/files.ts:9]: Actually -- rather than write this ourselves, what do you think of using something like https://www.npmjs.com/package/pretty-bytes (11m weekly downloads)

File: src/components/BulkUploadListItem.tsx
Comment (slifty) [inline src/components/BulkUploadListItem.tsx:23]: w00t file size is back!

Rather than having all these commits with thrash would it be easy to reorder / smoosh / remove things so that it's just introduced once?

**Review by slifty — CHANGES_REQUESTED** [review]
Just did a first pass code review.  Happy to talk through any of this / not trying to blow up progress with some of the more nit picky comments.

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx:98]: It does not, unfortunately. I could probably refactor it to have that by making `metadata` an object instead of an array, so we can use the object keys as part of the key construction… I'll try that!

File: src/pages/AddData.tsx
Comment (reefdog) [reply src/pages/AddData.tsx:13]: I hear you, but this is acceptable in the history of the project. This is just moving a string that already existed in the `BulkUploader` component — which we are deleting — inline to this template. Jason's uploader work, which will be the next PR that I open, will replace it with the real uploader.

File: src/utils/datetime.ts
Comment (reefdog) [reply src/utils/datetime.ts:32]: I sometimes (though rarely) have utility functions (or other variable creation) inside these utility files that _aren't_ meant for export, just some internal syntactic sugar. So an advantage of the manual export at the end is a single manifest that lets the developer eyeball all the exports from a file. Just IMO.

File: src/components/BulkUploadList.tsx
Comment (reefdog) [reply src/components/BulkUploadList.tsx]: Done by a following commit, but history has now been rewritten regardless so this was never even there! 💨

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx]: > I just realized these utilities don't take inputs / are not true functions.
>
> Can we adjust this and the FileSize function so they take input, humanize, and return the humanized output?

Refactor has removed the "not a true function" bit; they're both now just variable assignments (and any prerequisite do functions take input).

> Also, what do you think of removing humanized from the variable names (I'm not feeling strongly on this, but the prefix smelled a little of hungarian notation on first read)

If I removed the `humanized` prefix, `timestamp` will be fine but `fileSize` is already assigned from the prop destructuring, so we'd need to find some other way of indicating it. I could change the `humanized` prefix to something like `pretty` or `formatted`, but those feel a little like just changing the lipstick on the pig if you're hoping to fully avoid Hungarian.

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx]: I think that's a good idea! We iterated here from a previous version where there was less switch-able logic going on, but with where we've landed, a switch seems best.

File: src/utils/datetime.ts
Comment (reefdog) [reply src/utils/datetime.ts:1]: I actually thought best practice was the opposite: add it in its own commit, but bundle them together in a PR. But after further thought I'm not sure why that really helps, and also it's more difficult to reconstruct history from the basic git log/blame tools.

So, wish granted!

File: src/components/BulkUploadListItem.tsx
Comment (slifty) [reply src/components/BulkUploadListItem.tsx]: @reefdog I'm OK with it as is, given the context!

File: src/utils/datetime.ts
Comment (slifty) [reply src/utils/datetime.ts:1]: I probably asked you to do it the other way in a past project; oh to have been so young and naive.

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx]: Will do!

**Comment by reefdog** [comment]
@slifty Success! I've grouped the commits more sensibly (thank you for that prompt) and addressed what I believe were all your concerns. Can you give it another review? (It won't let me re-request one from you. 😵.)

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx]: @slifty I at least changed `humanized*` (which is weird, subjective) to `relativeTimestamp` and `formattedFileSize`.

File: src/utils/datetime.ts
Comment (reefdog) [reply src/utils/datetime.ts:1]: Probably, it does sound like you. With your endless requests and rowing.

File: package.json
Comment (slifty) [inline package.json:9]: Could you put this in its own commit?

File: src/components/BulkUploadListItem.tsx
Comment (slifty) [inline src/components/BulkUploadListItem.tsx:18]: Huh, I'm used to enum names (not values) being ALL_CAPS but I suppose if the linter doesn't care then why should we!  This is not a change request per se, just a musing.

File: src/components/BulkUploadListItem.tsx
Comment (slifty) [inline src/components/BulkUploadListItem.tsx:99]: this raises a deep philosophical question: if something has one space on either side of it, is it padded by two spaces or is it padded by one space.  If two: can something be padded by one space?  If one: are we saying the spaces are the same space?

**Review by slifty — APPROVED** [review]
HOORAY!

I left you some really useful final feedback.

(please smash the merge button at your leisure)

File: package.json
Comment (reefdog) [reply package.json:9]: `:twitch:`

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx:18]: Ooh, actually, I do like that. Linter doesn't care but I'll change it anyway.

File: src/components/BulkUploadListItem.tsx
Comment (reefdog) [reply src/components/BulkUploadListItem.tsx:99]: Superficially, you're right, and I can illustrate it with a CSS example: we'd never say "padded by 20px" to mean the sum of padding. We would mean 20px in each relevant dimension. E.g., a box with 5px of padding on each side would be "padded by 5px", not padded by 20px. (If it were padded in only one dimension, say, horizontally but not vertically, then either context or explanation would make that clear.)

However, this is a _bullet_, which has only one external surface. Therefore to describe its padding, we _do_ sum the values of "both sides" — that is, spacing off any axis. And so one space on either "side" is really two spaces distributed symmetrically off the horizontal poles of the unisurface. I thought it easier to say "padded by two spaces".

Thank you for coming to my TED talk.

Comment (reefdog) [reply src/components/BulkUploadListItem.tsx:99]: That was offensively clear. Or rather, so clear that your explanation is offensive to me.

**Pull request merged**
